### PR TITLE
Remove matplotlib, plotly, scipy, and scikit-learn dependencies

### PR DESCRIPTION
**Pull request by veered**
The use of `openai.embeddings_utils` adds some large dependencies. The actual functions being used from that package don't require any of those dependencies. 

I ported over the two functions from the [original file](https://github.com/openai/openai-python/blob/main/openai/embeddings_utils.py).

### CODE REVIEW

File: gpt_index/embeddings/base.py [hunk 1]
```diff
@@ -22,4 +22,6 @@ def get_text_embedding(self, text: str) -> List[float]:
 
     def similarity(self, embedding1: EMB_TYPE, embedding2: EMB_TYPE) -> float:
         """Get embedding similarity."""
-        return cosine_similarity(embedding1, embedding2)
+        product = np.dot(embedding1, embedding2)
+        norm = np.linalg.norm(embedding1) * np.linalg.norm(embedding2)
```
Comment (veered) [inline gpt_index/embeddings/base.py:26]: According to the OpenAI API docs, the embeddings are already normalized so this shouldn't be necessary.

Maybe the old embeddings weren't normalized, I'm not sure.

File: gpt_index/embeddings/openai.py [hunk 2]
```diff
@@ -3,7 +3,8 @@
 from enum import Enum
 from typing import List
 
-from openai.embeddings_utils import get_embedding
+import openai
+from tenacity import retry, stop_after_attempt, wait_random_exponential
```
Comment (veered) [inline gpt_index/embeddings/openai.py:7]: I don't know if you actually want to keep all the retry logic, but that's what they have in original code.

File: setup.py [hunk 3]
```diff
@@ -16,11 +16,8 @@
     "dataclasses_json",
     "transformers",
     "nltk",
-    # for openAI embeddings
-    "matplotlib",
-    "plotly",
-    "scipy",
-    "scikit-learn",
+    "numpy",
```
Comment (veered) [inline setup.py:19]: `openai` already depends on this, so it isn't a new dependency

File: setup.py [hunk 4]
```diff
@@ -16,11 +16,8 @@
...
+    "tenacity",
```
Comment (veered) [inline setup.py:20]: If you don't want to keep the retry logic, this can be removed

File: gpt_index/embeddings/base.py (hunk 1 above)
Comment (jerryjliu) [reply gpt_index/embeddings/base.py:26]: probably not a bad idea to explicitly compute it here just to be safe

File: gpt_index/embeddings/openai.py [hunk 5]
```diff
@@ -69,6 +70,15 @@ class OpenAIEmbeddingMode(str, Enum):
 }
 
 
+@retry(wait=wait_random_exponential(min=1, max=20), stop=stop_after_attempt(6))
+def get_embedding(
```
Comment (jerryjliu) [inline gpt_index/embeddings/openai.py:74]: This works. Would you be able to add a comment (or a docstring) referencing the original [file link](https://github.com/openai/openai-python/blob/main/openai/embeddings_utils.py)? Just so we know it's copied.

**Review by jerryjliu — COMMENTED** [review]
Thanks for doing this PR! 🙏 This is super useful for reducing num dependencies

**Pull request merged**